runtime/lua/vim/filetype.lua

```diff
@@ -2377,7 +2362,21 @@ local function dispatch(ft, path, bufnr, ...)
 end
 
 ---@private
-local function match_pattern(name, path, tail, pat)
+local function match_pattern(name, path, tail, pat, expand_env)
+  if expand_env then
+    local return_early
+    pat = pat:gsub('%${(%S-)}', function(env)
+      -- If an environment variable is present in the pattern but not set, there is no match
+      if not vim.env[env] then
+        return_early = true
+        return nil
+      end
+      return vim.env[env]
+    end)
+    if return_early then
+      return false
+    end
```

Comment:
Nice improvement but why is `nsub` `2` and not `0` in this case?

`lua =("${abc}/${def}"):gsub("%${(%S-)}", function() return false end)`

Reply:
I guess I'm just confused because it says `gsub also returns, as its second value, the total number of substitutions made.` in the docs and not the total number of matches.

Reply:
>I guess I'm just confused because it says gsub also returns, as its second value, the total number of substitutions made. in the docs and not the total number of matches.

You're right, `:h luaref.txt` does say this. But in https://www.lua.org/manual/5.1/manual.html#pdf-string.gsub it says the second argument is the total number of matches.

I am not sure where the discrepancy comes from, but given that I don't think my suggestion is valid.